src/kbc_modules/cc_kbc/mod.rs

```diff
@@ -94,4 +127,116 @@ impl Kbc {
             .ok_or_else(|| anyhow!("TEE rsa key missing"))?;
         response.decrypt_output(key)
     }
+
+    fn tee(&self) -> &str {
+        &self.tee
+    }
+
+    fn kbs_uri(&self) -> &str {
+        &self.kbs_uri
+    }
+
+    fn http_client(&mut self) -> &mut reqwest::Client {
+        &mut self.http_client
+    }
+
+    async fn establish_kbs_session(&mut self) -> Result<()> {
+        let kbs_uri = self.kbs_uri().to_string();
+        let parsed_kbs_uri =
+            Url::parse(&kbs_uri).map_err(|e| anyhow!("Parse KBS URL Failed: {:?}", e))?;
+        let auth_request = Request::new(self.tee().to_string());
+        let auth_response = self
+            .http_client()
+            .post(format!("{}/auth", kbs_uri))
+            .header("Content-Type", "application/json")
+            .body(serde_json::to_string(&auth_request)?)
+            .send()
+            .await?;
+
+        let mut cookies = auth_response
+            .headers()
+            .get_all(reqwest::header::SET_COOKIE)
+            .iter()
+            .peekable();
+        if cookies.peek().is_some() {
+            self.cookie_store.set_cookies(&mut cookies, &parsed_kbs_uri);
+        }
+
+        let challenge = auth_response.json::<Challenge>().await?;
+        self.nonce = challenge.nonce.clone();
+
+        let attestation_evidence = self.generate_evidence()?;
+        let attest_response = self
+            .http_client()
+            .post(format!("{}/attest", kbs_uri))
+            .header("Content-Type", "application/json")
+            .header(
+                reqwest::header::COOKIE,
+                self.cookie_store
+                    .cookies(&parsed_kbs_uri)
+                    .ok_or_else(|| anyhow!("Authenticated but cookie_store is still empty."))?,
+            )
+            .body(serde_json::to_string(&attestation_evidence)?)
+            .send()
+            .await?;
+        if attest_response.status() != reqwest::StatusCode::OK {
+            // TODO: Parse Error description payload
+            bail!("KBS attest failed")
+        }
+
+        Ok(())
+    }
+
+    async fn request_kbs_resource(&mut self, resource_url: String) -> Result<Response> {
+        let parsed_kbs_uri =
+            Url::parse(self.kbs_uri()).map_err(|e| anyhow!("Parse KBS URL Failed: {:?}", e))?;
+
+        if self.cookie_store.cookies(&parsed_kbs_uri).is_none() {
+            self.establish_kbs_session().await?;
+        }
```

Comment:
I know it was my suggestion to check the cookie store, but maybe instead of keeping our own cookies store and handling the set-cookie header manually, we can revert to the original implementation (with built-in private cookie store), and just keep a flag `bool is_authenticated`. It starts as false, and after successful `/attest` it is changed to true. If we ever get UNAUTHORIZED we changed it to false again.

Don't you think it's simpler than the whole handling of cookie_store ?